LG-3850: Implement Mobile Navigation

aduth opened the pull request:
Implements mobile navigation site-wide, with distinct navigation for help section.

**Live Preview:** https://federalist-17bd62cc-77b7-4687-9c62-39b462ce6fd5.app.cloud.gov/preview/18f/identity-site/aduth-lg-3850-mobile-nav/

**Screenshots:**

Main|Help Section
---|---
![Simulator Screen Shot - iPhone 12 Pro - 2021-01-06 at 14 46 35](https://user-images.githubusercontent.com/1779930/103813441-14a4e280-502e-11eb-9e09-c0f42b144e08.png)|![Simulator Screen Shot - iPhone 12 Pro - 2021-01-06 at 14 34 29](https://user-images.githubusercontent.com/1779930/103812302-53d23400-502c-11eb-9873-f15874cb6308.png)

**Note:** In the above screenshots, the rounded corners on the search field are not expected. This will be fixed separately by #515.

aduth commented:
This is ready for review.

zachmargolis commented on _sass/components/_btn.scss:
```diff
@@ -45,7 +61,7 @@
 
 @include at-media-max("desktop") {
-  .btn-lang-toggle.mobile {
+  div.btn-lang-toggle {
```
when do we use this toggle on something that isn't a `<div>`

aduth replied:
> when do we use this toggle on something that isn't a `<div>`

`btn-lang-toggle` is an existing (albeit not-so-accurate) class name used for both the static label + list (mobile) and button + dropdown (desktop) language picker variations:

https://github.com/18F/identity-site/blob/83943a8a6a8747c1f15ec1809c48a45c3a6128db/_includes/header.html#L45-L50

![image](https://user-images.githubusercontent.com/1779930/103814642-0788f300-5030-11eb-8e3b-979392204b4f.png)

https://github.com/18F/identity-site/blob/83943a8a6a8747c1f15ec1809c48a45c3a6128db/_includes/language_search.html#L16-L21

![image](https://user-images.githubusercontent.com/1779930/103814653-0d7ed400-5030-11eb-823c-c317c76c099d.png)

Because I dropped the `.mobile` class here, `div` was the next best thing to assign styles for the existing element.

Separately, I'd think we could probably rename this to something which describes both cases equally well if we wanted to share styles. Something like `language-picker__label`.

zachmargolis replied:
ok! I think the `.mobile` class makes the intent clearer? But this is not my wheelhouse

aduth replied:
> ok! I think the `.mobile` class makes the intent clearer? But this is not my wheelhouse

Yeah, I'd intended to try to refactor this a bit to treat it more as "list form" vs. "dropdown form" (as opposed to "desktop" vs. "mobile"), but in any case, a class would likely be clearer. I'll plan to make an update.

zachmargolis commented on _i18n/fr.yml:
```diff
@@ -308,54 +287,53 @@ help_pages:
-      * Vous réalisez que vous avez plusieurs comptes avec différentes adresses de courriel et vous souhaitez avoir un seul compte. 
-      * Vous ne pouvez plus accéder à votre compte et vous n'avez plus accès à vos méthodes d'authentification. 
-      * Vous n'avez plus besoin d'avoir un compte login.gov 
-      
```
🙏 thank you for getting all these trailing whitespaces

zachmargolis commented on _layouts/help_landing.html:
```diff
@@ -7,7 +7,7 @@
   {% include skip_nav.html %}
   {% include banner.html %}
-  {% include language_search.html logo=true signin=false contact=true %}
+  {% include header--help.html %}
```
Our other partials use underscores, what was the motivation for dashes?

aduth replied:
> Our other partials use underscores, what was the motivation for dashes?

I'm not especially attached to it, but there was one example of precedent here:

https://github.com/18F/identity-site/blob/demo/_includes/hero--search.html

As I understand and applied it here, the idea for it could be as a [BEM](http://getbem.com/)-like modifier, as this is intended to serve as an in-place variant substitute for the base `header.html` partial.

I'm not sure how often these sorts of cases would arise, since in many we could simply parameterize the partial to change its behavior. In this case, since there's enough differences between the two variants, it probably would have been more work to handle the logic for parameters than to create an entirely different partial. That being said, I'd not want to set a convention here if it's not something we generally want to encourage.

zachmargolis replied:
ok! As long as this it's not the only one, I guess that's fine?

aduth replied:
Thanks for the clarification @cannandev ! I agree that we can simplify a bit to remove `language_search.html`. Per your suggestion, I did this in 7539b80.

zachmargolis reviewed (approved):
LGTM

nickttng commented:
@shawnique (and @zachmargolis since you're familiar with this issue) 

For the mobile menu - does either of you have any concerns with the placement of `Sign in with...` button, which is under the `Search` component? This might create a risk of users entering their personal info in the search field (like what we've seen in the past). And although the 'Search' button has text, it might not be obvious enough before users.

Potential ways to reduce the risk, so that both the `Search` component and `Sign in with...` button is not near each other:

- Move the `Search` component under the `Contact us` text if `Sign in with...` is a priority at the top
- Move the `Sign in with...` button under the "Help center" link item if `Search` is a priority at the top

I lean toward the first option. How it would be ordered:

1. `Sign in with...` button
2. Link items
3. `Search` component
4. "Contact us"
5. Language options

zachmargolis commented:
> * Move the `Search` component under the `Contact us` text if `Sign in with...` is a priority at the top
> 
> I lean towards the first option. How it would be ordered similarly to desktop ('Sign in...' above search):

I agree, I think moving the search would minimize our chances for confusion, and still keep search functionality "above the fold"

nickttng commented:
@shawnique Per @shawnique in a [Slack convo](https://gsa-tts.slack.com/archives/C01710KMYUB/p1609967941493100) and Zach's comment above, let's move `Search` under the 'Contact us' text.

The order:

1. `Sign in with...` button
2. Link items
3. 'Contact us'
4. `Search` component
5. Language options

cannandev reviewed (changes requested):
Looks like we lost the active styling on current elements. Can we add that per design?
![image](https://user-images.githubusercontent.com/381122/103827542-44141900-5047-11eb-82e0-a9d96e4a89f9.png)

cannandev commented:
Also, the page `nav_sidenav` is still visible. Can you hide that on mobile?
![image](https://user-images.githubusercontent.com/381122/103832144-9e16dd80-504b-11eb-8a94-9c1f841d95a8.png)

aduth commented:
> Looks like we lost the active styling on current elements. Can we add that per design?

Just want to clarify what you mean by active styling here: Based on the Figma designs and screenshot you shared, I'm expecting there to be a blue, bolded effect on link for the current subpage in the help section, which is currently implemented:

![image](https://user-images.githubusercontent.com/1779930/103897581-cb05d780-50c1-11eb-8b82-e0a3ae9cfb92.png)

Re-referencing the design, I do see some additional styling which will need to be implemented for the `:active` (tapped / mousedown) links in the menu. I'll plan to implement that shortly.

![image](https://user-images.githubusercontent.com/1779930/103897747-ff799380-50c1-11eb-90f9-982d2e706270.png)

aduth commented:
From https://github.com/18F/identity-site/pull/511#discussion_r552951998:

> > ok! I think the `.mobile` class makes the intent clearer? But this is not my wheelhouse
> 
> Yeah, I'd intended to try to refactor this a bit to treat it more as "list form" vs. "dropdown form" (as opposed to "desktop" vs. "mobile"), but in any case, a class would likely be clearer. I'll plan to make an update.

Added distinguishing classes in 3ec24f7.

---

From https://github.com/18F/identity-site/pull/511#discussion_r553024811:

> @aduth I provided suggested code for 1). Can you include the other changes in this branch?

Eliminated `language_picker.html` partial in 7539b80.

---

From https://github.com/18F/identity-site/pull/511#issuecomment-755740215:

> @aduth Per @shawnique in a [Slack convo](https://gsa-tts.slack.com/archives/C01710KMYUB/p1609967941493100) and Zach's comment above, let's move `Search` under the 'Contact us' text.

Rearranged mobile navigation in b74ff61.

---

From [Slack](https://gsa-tts.slack.com/archives/C01710KMYUB/p1609963576489400?thread_ts=1609960081.482500&cid=C01710KMYUB):

>Although I’m glad I just took that screenshot, since I can see a little regression with the search button label 😅 

Fixed in 282aee4.

---

From https://github.com/18F/identity-site/pull/511#issuecomment-756115606:

> Re-referencing the design, I do see some additional styling which will need to be implemented for the `:active` (tapped / mousedown) links in the menu. I'll plan to implement that shortly.

Added focus styling in ed0c04d.

---

From https://github.com/18F/identity-site/pull/511#issuecomment-755774199:

> Also, the page `nav_sidenav` is still visible. Can you hide that on mobile?

Fixed mobile side navigation in abf7080. I believe this might also have surfaced as part of the testing session yesterday.

aduth commented on _includes/header.html:
```diff
@@ -1,20 +1,21 @@
 <div class="usa-header usa-header--extended">
```
Incorporated into broader commit 7539b80.

nickttng commented:
Not critical, but to note for the future to improve - Search component makes it clear that it's not right-sized (see screenshots); however, in the bigger picture, we could consider setting a max-width for the mobile menu, like the USWDS website (see example)


## Brochure site

![search-ipad-768](https://user-images.githubusercontent.com/6327082/104042954-b8160480-51a0-11eb-8210-233a5e2c4c27.png)
![search-surface-720](https://user-images.githubusercontent.com/6327082/104042957-b8ae9b00-51a0-11eb-9c82-c5ea440510b6.png)

## USWDS site

![uswds-mobile-menu](https://user-images.githubusercontent.com/6327082/104043427-6cb02600-51a1-11eb-84f2-18514eae5253.png)

aduth commented:
> however, in the bigger picture, we could consider setting a max-width for the mobile menu, like the USWDS website (see example)

It appears that at some point in the past we set an explicit override width of `85vw` to the mobile navigation. I removed this in a42f2e7 so that the USWDS default takes effect once more.

![image](https://user-images.githubusercontent.com/1779930/104045680-ea2d6380-51ac-11eb-903f-db5dd9234a80.png)

aduth commented:
Oops, I see in the screenshot now we're wrapping the "Sign In" button. I suppose we don't want that. Let me see if there's a happy medium.

aduth commented:
> Oops, I see in the screenshot now we're wrapping the "Sign In" button. I suppose we don't want that. Let me see if there's a happy medium.

Fits after removing button's margins (2eb5bf2):

![image](https://user-images.githubusercontent.com/1779930/104048675-a25d0b00-51b1-11eb-84b4-79078f45d901.png)

Offset this for desktop in 21433be.

aduth commented:
Going to go ahead and merge this as-is, but let me know if you spot any other defects @nickttng and I'll put together a follow-up pull request to address those.
